### DISCUSSION TITLE
Attempting to use `Distributed` and consistently getting a `MethodError` on `sorted_deppairs`

### DISCUSSION THREAD

**Discussion by jmuchovej** [JuliaPOMDP/POMDPs.jl#474] [Debugging Help] [answered]
When attempting to solve many POMDPs in parallel (each process is a POMDP-solver pair, no communication between pairs is needed), I consistently run in to an issue where `sorted_deppairs` cannot be dispatched, yet when I run the same functions on the main threads, everything proceeds accordingly.

The error, in brief:
```
On worker 6:
MethodError: no method matching sorted_deppairs(::Type{BoxWorld}, ::NTuple{4, Symbol})
Closest candidates are:
  sorted_deppairs(::Type{<:POMDP}, ::Any) at [~/.local/conda/lib/julia/packages/POMDPs/XBTe5/src/gen_impl.jl:66](https://file+.vscode-resource.vscode-cdn.net/Users/john/10-projects/10.01-agent-programs/agent-programs/_research/~/.local/conda/lib/julia/packages/POMDPs/XBTe5/src/gen_impl.jl:66)
  sorted_deppairs(::Type{<:MDP}, ::Any) at [~/.local/conda/lib/julia/packages/POMDPs/XBTe5/src/gen_impl.jl:56](https://file+.vscode-resource.vscode-cdn.net/Users/john/10-projects/10.01-agent-programs/agent-programs/_research/~/.local/conda/lib/julia/packages/POMDPs/XBTe5/src/gen_impl.jl:56)
  sorted_deppairs(::Dict{Symbol, Vector{Symbol}}, ::Any) at [~/.local/conda/lib/julia/packages/POMDPs/XBTe5/src/gen_impl.jl:77](https://file+.vscode-resource.vscode-cdn.net/Users/john/10-projects/10.01-agent-programs/agent-programs/_research/~/.local/conda/lib/julia/packages/POMDPs/XBTe5/src/gen_impl.jl:77)
Stacktrace:
...
```

`BoxWorld` is a `POMDP`, as evidence:
```diff
       From worker 7:	typeof(pomdp) = BoxWorld
       From worker 7:	typeof(pomdp) <: POMDP = true
       From worker 8:	typeof(pomdp) = BoxWorld
       From worker 8:	typeof(pomdp) <: POMDP = true
       From worker 4:	typeof(pomdp) = BoxWorld
       From worker 4:	typeof(pomdp) <: POMDP = true
       From worker 2:	typeof(pomdp) = BoxWorld
       From worker 2:	typeof(pomdp) <: POMDP = true
       From worker 3:	typeof(pomdp) = BoxWorld
       From worker 3:	typeof(pomdp) <: POMDP = true
       From worker 5:	typeof(pomdp) = BoxWorld
       From worker 5:	typeof(pomdp) <: POMDP = true
+      From worker 6:	typeof(pomdp) = BoxWorld
+      From worker 6:	typeof(pomdp) <: POMDP = true
       From worker 9:	typeof(pomdp) = BoxWorld
       From worker 9:	typeof(pomdp) <: POMDP = true
```

What's most perplexing is that if I don't use `Distributed`, everything runs just fine.
<img width="695" alt="image" src="https://user-images.githubusercontent.com/5000729/232067888-20d71529-956f-4dc4-ac0d-c2acd601d679.png">

Potential causes for error: I'm using a Jupyter Notebook through VSCode, but I've also replicated the error on Jupyter Lab and in a shell script.

`do_null_program_sim` does some extra processing before running `do_pomdp_sim`, which is:
```julia
function do_pomdp_sim(
    seed::Int, items::Vector, boxes::Vector{Box}, target
)
    # @show seed
    Random.seed!(seed)

    program = @agent items begin end

    solver = QMDPSolver(;
        max_iterations=100,
        belres=1e-3,
        verbose=false
    )

    pomdp = FruitWorld(
        program;
        boxes=boxes,
        items=items,
        target=target
    )

    @show typeof(pomdp)
    @show typeof(pomdp) <: POMDP

    policy = solve(solver, pomdp)

    hr = HistoryRecorder(; max_steps=30)
    hist = simulate(hr, pomdp, policy)

    return (seed=seed, pomdp=pomdp, policy=policy, hist=hist)
end
```
`FruitWorld` and `BoxWorld` are synonymous (they need to be differently named for other reasons).

**Comment by zsunberg** [comment]
Well, this is perplexing and frustrating!

I don't know why the dispatch isn't working. It probably has something to do with genout (https://github.com/JuliaPOMDP/POMDPs.jl/blob/851535f2c2b9f08d160f69b4326f817c966f03fd/src/gen_impl.jl#L19) being a `@generated` function

**Reply by zsunberg** [reply]
Unfortunately I can't think of a simple workaround. I'll try to experiment with it today and see if I can reproduce.

**Reply by zsunberg** [reply]
If you are really wanting to get something working as fast as possible, you could try make a specific implementation of POMDPs.genout for your problem.

**Reply by zsunberg** [reply]
@jmuchovej I tried to reproduce the error with this code:
```julia
procs = addprocs(2)
@everywhere using POMDPs, Random
@everywhere begin
    struct DistributedDeterministic{T}
        x::T
    end
    Base.rand(rng::AbstractRNG, d::DistributedDeterministic) = d.x

    struct DistributedTestMDP <: MDP{Int, Int} end
    POMDPs.transition(::DistributedTestMDP, s, a) = DistributedDeterministic(s+a)
    m = DistributedTestMDP()
    @gen(:sp)(m, 1, 1)
end
rmprocs(procs)
```
Unfortunately, this code executed without error. Can you either share all your code or help me find the differences between this minimal example and your code that results in an error.

**Reply by zsunberg** [reply]
Also it's probably worth checking which version of julia you are using and trying it with the latest just in case this is a bug in julia that has been fixed.

**Reply by jmuchovej** [reply, marked as answer]
Thanks for testing this out! 🙂

While looking into this issue – it seems like another library I've been using _might_ be causing a problem (how still remains unclear).

I'm using `DrWatson` to do stuff like ensure we're using the right Julia project environment and the like, but for whatever reason when it's paired with `Distributed` things start "acting out". 

- https://github.com/JuliaDynamics/DrWatson.jl/issues/269
- https://github.com/JuliaDynamics/DrWatson.jl/issues/270

Definitely not a satisfying answer, but once I swapped out `DrWatson` for traditional Julia `Pkg.activate(...)` commands, **everything seems to be working correctly**, so I'll go ahead and mark this as answered and hopefully this is helpful to anyone else trying to use the combo. 😅
